pravega/pravega
# Issue 5859: (v0.9.1) Cherry-pick various SegmentStore bug fixes

**Change log description**  
Cherry-picking these PRs:
- #5841: Issue #5840: (SegmentStore) Fixed a deadlock in SegmentKeyCache.
- #5851: Issue #5850: (SegmentStore) Fixed a bug in WriterTableProcessor where it would attempt to flush to a deleted segment.
- #5586: Issue #5581: (SegmentStore) Disabling non-essential cache inserts if cache utilization is high 
- #5804: Issue #5789: (SegmentStore) Improving stability during Segment Container Recoveries
- #5811: Issue #5810: (SegmentStore) Fixed a StorageWriter bug that could lead to data loss 
- #5783: Issue #5771: (SegmentStore) Reducing the amount of heap memory used when doing Table Segment Reads.

**Purpose of the change**  
Fixes #5859.

**What the code does**  
Cherry-pick.

**How to verify it**  
Build must pass. Verify intended commits have been cherry-picked into this branch.
